src/librustc_trans/context.rs

```diff
@@ -322,6 +355,11 @@ unsafe fn create_context_and_module(sess: &Session, mod_name: &str) -> (ContextR
     let llvm_target = sess.target.target.llvm_target.as_bytes();
     let llvm_target = CString::new(llvm_target).unwrap();
     llvm::LLVMRustSetNormalizedTarget(llmod, llvm_target.as_ptr());
+
+    if is_pie_binary(sess) {
+        llvm::LLVMRustSetModulePIELevel(llmod);
```

Comment:
You might need to set the `PICLevel` too, even if the relocation model is `PIC`.
One potentially related failure is:
```
note: ld: illegal text-relocation to '_panic_loc21555' in i686-apple-darwin/stage1/lib/rustlib/i686-apple-darwin/lib/std-f53fb285.0.o from '__ZN103_$LT$core..slice..RSplitN$LT$$u27$a$C$$u20$T$C$$u20$P$GT$$u20$as$u20$core..iter..iterator..Iterator$GT$4next17he6277dfc64030327E' in i686-apple-darwin/stage1/lib/rustlib/i686-apple-darwin/lib/std-f53fb285.0.o for architecture i386
```